File: samples/hemisphere_gen.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <glm/glm.hpp>

std::vector<glm::vec3> sphere_gen(float radius, int num_points, int num_steps, glm::vec3 origin = glm::vec3(0.0f, 0.0f, 0.0f)) {
    std::vector<float> radius_in_each_step(num_steps);
    for(int i = 0; i < num_steps; i++) {
        radius_in_each_step[i] = sin((M_PI / 2) * (static_cast<float>(i) / num_steps));
    }
    
    float total = 0;
    for(auto val : radius_in_each_step) {
        total += val;
    }

    std::vector<int> n_view_in_each_steps(num_steps);
    for(int i = 0; i < num_steps; i++) {
        n_view_in_each_steps[i] = round(num_points * (radius_in_each_step[i] / total));
    }
    
    std::vector<glm::vec3> points;
    for(int idx = 0; idx < num_steps; idx++) {
        for(int i = 0; i < n_view_in_each_steps[idx]; i++) {
            float theta = radius_in_each_step[idx];
            float phi = 2 * M_PI * i / n_view_in_each_steps[idx];
            float x = radius * sin(theta) * cos(phi) + origin.x;
            float y = radius * sin(theta) * sin(phi) + origin.y;
            float z = radius * cos(theta) + origin.z;
            points.push_back(glm::vec3(x, y, z));
        }
    }
    return points;
}



int main() {
    float radius = 1.5f;
    int num_points = 100;
    int num_steps = 5;
    glm::vec3 origin(0.0f, 0.0f, 0.0f);

    std::vector<glm::vec3> points = sphere_gen(radius, num_points, num_steps, origin);

    std::cout << "Generated points: \n";
    for (auto& point : points) {
        std::cout << "(" << point.x << ", " << point.y << ", " << point.z << ")\n";
    }

    return 0;
}

File: samples/hemisphere_gen.py
import numpy as np
from scipy.spatial.transform import Rotation as R
import open3d as o3d

def theta_phi_to_xyz(theta, phi, radius):
    """
    convert theta, phi to xyz
    """
    x = radius * np.sin(theta) * np.cos(phi)
    y = radius * np.sin(theta) * np.sin(phi)
    z = radius * np.cos(theta)

    return np.array([x, y, z])

def sphere_gen(radius: float,
               num_points: int, 
               num_steps: int,
               origin: np.ndarray = np.array([0, 0, 0], dtype=np.float32) 
               ) -> list:
    """
    generate hemisphere point cloud
    """
    # step 1. set points num in each height as n_view
    thetas = [(np.pi/2) * (i / num_steps) for i in range(num_steps)]
    radiuses = np.sin(thetas)
    n_view = np.round(num_points * radiuses / np.sum(radiuses)).astype(int)

    # step 2. generate points
    points = []
    rmats = []

    for n_v, theta in zip(n_view, thetas):
        for i in range(n_v):
            phi = 2 * np.pi * i / n_v
            pts = theta_phi_to_xyz(theta, phi, radius) + origin

            # calculate orientation
            u_z = -1 * theta_phi_to_xyz(theta, phi, 1)
            u_y = theta_phi_to_xyz(theta + np.pi/2, phi, 1)
            u_x = np.cross(u_y, u_z)

            rmat = np.linalg.inv(np.array([u_x, u_y, u_z]))

            points.append(pts)
            rmats.append(rmat)

    transforms = []

    # points and oris to transform matrix
    for point, rmat in zip(points, rmats):
        transform = np.eye(4)
        transform[:3, 3] = point
        transform[:3, :3] = rmat
        transforms.append(transform)

    return transforms

# Open3D를 사용하여 3D 점군을 시각화합니다.
def visualize_cloud(transforms):
    meshes = []

    for tf in transforms:
        mesh = o3d.geometry.TriangleMesh.create_coordinate_frame()
        meshes.append(mesh.transform(tf).scale(0.1, [0, 0, 0]))
    
    o3d.visualization.draw_geometries(meshes)


if __name__ == "__main__":
    points = sphere_gen(20, 100, 5)

    visualize_cloud(points)